syntax = "proto3";

option java_multiple_files = true;
option java_package = "newsletter";
option java_outer_classname = "NewsletterProto";

package newsletter;

service NewsletterService {
    rpc Subscribe(SubscriptionRequest) returns (SubscriptionResponse) {}
    rpc Unsubscribe(UserRequest) returns (SubscriptionResponse) {}
    rpc GetNotifications(UserRequest) returns (stream EventNotification) {}
}

message UserRequest {
    int32 userId = 1;
}

message SubscriptionRequest {
    int32 userId = 1;
    string city = 2;
    repeated string tags = 3;
}

message SubscriptionResponse {
    bool success = 1;
    string message = 2;
}

message EventNotification {
    enum NotificationType {
        NEW_EVENT = 0;
        EVENT_UPDATED = 1;
        EVENT_CANCELLED = 2;
    }
    NotificationType type = 2;
    string title = 3;
    string description = 4;
    string city = 5;
    repeated string tags = 6;
}